java/src/Ice/src/main/java/com/zeroc/IceInternal/Instance.java

```diff
@@ -761,10 +763,22 @@ public Class<?> apply(String typeId)
             if(pos != -1)
             {
                 String topLevelModule = typeId.substring(2, pos);
-                String pkg = _initData.properties.getProperty("Ice.Package." + topLevelModule);
-                if(pkg.length() > 0)
+                String[] packages = _builtInModulePackages.get(topLevelModule);
+                if (packages == null)
                 {
-                    c = getConcreteClass(pkg + "." + className);
+                    packages = _initData.properties.getPropertyAsListWithDefault("Ice.Package." + topLevelModule, null);
+                }
+
+                if(packages != null)
+                {
+                    for(String pkg : packages)
+                    {
+                        c = getConcreteClass(pkg + "." + className);
+                        if(c != null)
```

Comment:
If we add an annotation for Slice TypeIDs we can check here that the class has the expected type ID and fail if the TypeID doesn't match, or keep going if the class lacks the annotation.